src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -0,0 +1,187 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    internal class TdsWriter
+    {
+        private readonly TdsWriteStream _writeStream;
+
+        // Header length constants
+        private const int OutputHeaderLen = TdsEnums.HEADER_LEN;
+
+        // Out buffer variables
+        private readonly byte[] _outBuff; // internal write buffer - initialize on login
+        private int _outBytesUsed = OutputHeaderLen; // number of bytes used in internal write buffer - initialize past header
+
+        /// <summary>
+        /// Instantiate TdsWriter with a TdsWriteStream
+        /// </summary>
+        /// <param name="stream">A Tds Writer stream to work with</param>
+        /// <exception cref="ArgumentNullException"></exception>
+        public TdsWriter(TdsWriteStream stream)
+        {
+            _writeStream = stream ?? throw new ArgumentNullException(nameof(stream));
+            _outBuff = new byte[OutputHeaderLen]; // Adjust the buffer size as needed
+        }
+
+        #region Public APIs
+
+        /// <summary>
+        /// Writes short value to out buffer
+        /// </summary>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Is Async caller method</param>
+        /// <param name="ct">Cancellation token</param>
+        public async Task WriteShortAsync(short v, bool isAsync, CancellationToken ct)
```

Comment:
Why not using CancellationToken cancellationToken = default. That's kinda the default naming for them.

Why v as the name of the value param? You throw exceptions with nameof, having a nicer name would be better imho.

Why Task? You have a sync path and the ``EnsureBufferSpaceAsync`` method has a sync path, so you might want to look at ValueTask, especially if at some point the Stream is replaced with something modern, like Pipeline.

Reply:
ValueTask is the right choice because the underlying TdsStream also buffers the data before sending out to the network. So the ValueTask overloads of TdsStream can benefit here.

Reply:
+1 for ValueTask. It'll also avoid a heap allocation in sync cases